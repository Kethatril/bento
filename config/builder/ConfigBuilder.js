const { Config } = require("webpack-config"),
    parse = require("./modules/parse");

module.exports = class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    constructor(options = {}) {
        this._options = parse(options);
        this._config = new Config();
        this._modules = [];
        this._manifest = {};
    }

    use(...modules) {
        this._modules = this._modules.concat(modules);
        return this;
    }

    build(fn) {
        this._modules.push(fn);

        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._modules,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            let config = new Config(),
                options = this._options;

            try {
                (function next(err) {
                    let async;

                    // Increment counter variable and skip any indices that
                    // don't exist. This allows sparse arrays to be iterated.
                    do {
                        ++i;
                    } while (!(i in modules) && i !== len);

                    if (err === false || err) return reject(err);

                    if (i === len) return resolve(config.toObject());

                    let result;

                    result = modules[i].call(
                        {
                            async() {
                                async = true;
                                return next;
                            }
                        },
                        config,
                        options
                    );
                    if (!async) {
                        next(result);
                    }
                })();
            } catch (e) {
                reject(e);
            }
        });
    }
};
