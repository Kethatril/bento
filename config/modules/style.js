const { config, read } = require("../utils");
const ExtractTextPlugin = require("extract-text-webpack-plugin");

module.exports = config(instance => {
    return instance.merge({
        module: {
            rules: [
                {
                    test: /\.css$/,
                    include: /node_modules/,
                    use: ExtractTextPlugin.extract({
                        fallback: "style-loader",
                        publicPath: "../",
                        use: [
                            {
                                loader: "css-loader",
                                options: {
                                    sourceMap: read("enabled.sourceMap")
                                }
                            }
                        ]
                    })
                },

                {
                    test: /\.scss$/,
                    include: read("paths.input"),
                    use: ExtractTextPlugin.extract({
                        fallback: "style-loader",
                        publicPath: "../",
                        use: [
                            {
                                loader: "css-loader",
                                options: {
                                    sourceMap: read("enabled.sourceMap")
                                }
                            },

                            {
                                loader: "postcss-loader",
                                options: {
                                    ident: "postcss",
                                    plugins: () => {
                                        const browsers = read(
                                            "browserslist.browsers"
                                        );
                                        return [
                                            require("autoprefixer")({
                                                browsers
                                            })
                                        ];
                                    },
                                    sourceMap: read("enabled.sourceMap")
                                }
                            },
                            {
                                loader: "resolve-url-loader",
                                options: {
                                    sourceMap: read("enabled.sourceMap")
                                }
                            },
                            {
                                loader: "sass-loader",
                                options: {
                                    sourceMap: read("enabled.sourceMap")
                                }
                            }
                        ]
                    })
                }
            ]
        },

        plugins: [
            new ExtractTextPlugin({
                filename: `styles/${read("filename").replace(
                    "hash:",
                    "contenthash:"
                )}.css`,
                disable: !read("env.isProduction")
            })
        ]
    });
});
