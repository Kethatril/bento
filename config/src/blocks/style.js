import merge from "webpack-merge";
import ExtractTextPlugin from "extract-text-webpack-plugin";

export default function(config, options) {
    return merge({
        module: {
            rules: [
                {
                    test: /\.css$/,
                    include: /node_modules/,
                    use: ExtractTextPlugin.extract({
                        fallback: "style-loader",
                        publicPath: "../",
                        use: [
                            {
                                loader: "css-loader",
                                options: {
                                    sourceMap: options.get("enabled.sourceMap")
                                }
                            }
                        ]
                    })
                },

                {
                    test: /\.scss$/,
                    include: options.get("paths.input"),
                    use: ExtractTextPlugin.extract({
                        fallback: "style-loader",
                        publicPath: "../",
                        use: [
                            {
                                loader: "css-loader",
                                options: {
                                    sourceMap: options.get("enabled.sourceMap")
                                }
                            },

                            {
                                loader: "postcss-loader",
                                options: {
                                    ident: "postcss",
                                    plugins: () => {
                                        const browsers = options.get(
                                            "browserslist.browsers"
                                        );
                                        return [
                                            require("autoprefixer")({
                                                browsers
                                            })
                                        ];
                                    },
                                    sourceMap: options.get("enabled.sourceMap")
                                }
                            },
                            {
                                loader: "resolve-url-loader",
                                options: {
                                    sourceMap: options.get("enabled.sourceMap")
                                }
                            },
                            {
                                loader: "sass-loader",
                                options: {
                                    sourceMap: options.get("enabled.sourceMap")
                                }
                            }
                        ]
                    })
                }
            ]
        },

        plugins: [
            new ExtractTextPlugin({
                filename: `styles/${options
                    .get("filename")
                    .replace("hash:", "contenthash:")}.css`,
                disable: !options.get("env.isProduction")
            })
        ]
    });
}
