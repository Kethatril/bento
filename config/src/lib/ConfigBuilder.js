import { parse, resolve, select } from "~/lib/utils";
import ManifestKeeper from "~/lib/ManifestKeeper";

export default class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    manifest = new ManifestKeeper();

    constructor(options = {}) {
        this._options = parse(resolve(options));
        this._blocks = options.blocks.list || [];
        this._timeout = options.blocks.timeout || 3000;
    }

    get options() {
        return this._options;
    }

    build() {
        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._blocks,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            const timeout = this._timeout;

            let config = {},
                options = this._options,
                utils = { select: select(options), manifest: this.manifest };

            (function next(result) {
                let async;

                // Increment counter variable and skip any indices that
                // don't exist. This allows sparse arrays to be iterated.
                do {
                    ++i;
                } while (!(i in modules) && i !== len);

                if (result === false || result instanceof Error)
                    return reject(err);

                if (i === len) {
                    return resolve(result);
                }

                try {
                    result = modules[i](result, options, utils);
                } catch (e) {
                    return reject(e);
                }
                if (!result || typeof result !== "object") {
                    throw new Error(
                        "[InvalidValueError] Each block must either return a new config object or a promise that resolves a new config object."
                    );
                }

                if (result.then) {
                    const id = setTimeout(function() {
                        if (!async) return;
                        return reject(
                            new Error(
                                `[TimeoutError] ${timeout} ms exceeded. For async blocks, ensure the returned promise is resolved.`
                            )
                        );
                    }, timeout);

                    const done = function(result) {
                        clearTimeout(id);
                        next(result);
                    };

                    async = true;
                    result.then(done, done);
                }

                if (!async) {
                    next(result);
                }
            })(config);
        });
    }
}
