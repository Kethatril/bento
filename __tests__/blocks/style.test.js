import style from "~/blocks/style";
import ConfigBuilder from "~/lib/ConfigBuilder";

describe("blocks/style", () => {
    const options = {
        blocks: {
            babel: {},
            list: [style]
        }
    };

    it("matches snapshot", () => {
        const builder = ConfigBuilder.create(
            Object.assign(
                {},
                {
                    env: {
                        value: "development"
                    },

                    filename: "[name]"
                },
                options
            )
        );

        return builder.build().then(config => {
            expect(config).toMatchSnapshot();
        });
    });
});
