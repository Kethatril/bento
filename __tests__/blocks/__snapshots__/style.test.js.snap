// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blocks/style matches snapshot 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "include": /node_modules/,
        "test": /\\.css$/,
        "use": Array [
          Object {
            "loader": "<project>/node_modules/extract-text-webpack-plugin/dist/loader.js",
            "options": Object {
              "omit": 1,
              "publicPath": "../",
              "remove": true,
            },
          },
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": undefined,
            },
          },
        ],
      },
      Object {
        "include": undefined,
        "test": /\\.scss$/,
        "use": Array [
          Object {
            "loader": "<project>/node_modules/extract-text-webpack-plugin/dist/loader.js",
            "options": Object {
              "omit": 1,
              "publicPath": "../",
              "remove": true,
            },
          },
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": undefined,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "ident": "postcss",
              "plugins": [Function plugins],
              "sourceMap": undefined,
            },
          },
          Object {
            "loader": "resolve-url-loader",
            "options": Object {
              "sourceMap": undefined,
            },
          },
          Object {
            "loader": "sass-loader",
            "options": Object {
              "sourceMap": undefined,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    ExtractTextPlugin {
      "filename": "styles/[name].css",
      "id": 1,
      "options": Object {
        "disable": true,
      },
    },
  ],
}
`;
